File: medium/0122_best_time_to_buy_and_sell_stocks_ii/solution.cpp
#include <vector>

class Solution {
public:
  int maxProfit(std::vector<int> &prices) {
    int profit = 0;

    for (size_t i{}; i < prices.size(); ++i) {
      while (i < prices.size() - 1 && prices[i] > prices[i + 1])
        ++i;

      int buyPrice = prices[i];

      while (i < prices.size() - 1 && prices[i] < prices[i + 1])
        ++i;

      profit += prices[i] - buyPrice;
    }

    return profit;
  }
};

File: medium/0122_best_time_to_buy_and_sell_stocks_ii/solution.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0

        buy = 0
        while buy < len(prices) - 1:
            if prices[buy] >= prices[buy + 1]:
                buy += 1
                continue

            sell = buy + 1
            while sell < len(prices) - 1 and prices[sell] < prices[sell + 1]:
                sell += 1

            profit += prices[sell] - prices[buy]
            buy = sell

        return profit
